lib/perl/Genome/Sample.pm

```diff
@@ -267,11 +268,60 @@ sub sample_name_to_name_in_vcf {
 sub name_in_vcf {
     my $self = shift;
     my $sample_tcga_name = $self->extraction_label;
-    if ($sample_tcga_name and $sample_tcga_name =~ /^TCGA\-/) {
-        $self->debug_message("Found TCGA name: $sample_tcga_name for sample: %s", $self->name);
-        return $sample_tcga_name;
+
+    unless ($sample_tcga_name and $sample_tcga_name =~ /^TCGA\-/) {
+        $sample_tcga_name = $self->name;
+        my @tcga_names = $self->get_tcga_names;
+        if (@tcga_names and @tcga_names == 1) {
```

Comment:
There are cases that TCGA sample has multiple TCGA barcodes (from different sources) stored as external_name in sample attributes. It is ok.  If code reaches this step (no TCGA names as extraction_label and  @tcga_names != 1), the only way to find correct TCGA names is to parse TCGA submission data inventory .dat files. See comments in Genome::Info::TCGASpecialSampleNamingConversion.